snuba/datasets/plans/split_strategy.py

```diff
@@ -0,0 +1,40 @@
+from abc import ABC, abstractmethod
+from typing import Callable, Optional
+
+from snuba.clickhouse.query import Query
+from snuba.request.request_settings import RequestSettings
+from snuba.web import QueryResult
+
+SplitQueryRunner = Callable[[Query, RequestSettings], QueryResult]
+
+
+class QuerySplitStrategy(ABC):
+    """
+    Implements a query split algorithm. It works in a similar way as a
+    QueryExecutionStrategy, it takes a query, request settings and a query runner
+    and decides if it should split the query into more efficient parts.
+    If it can split the query, it uses the SplitQueryRunner to execute every chunk,
+    otherwise it returns None immediately.
+
+    The main difference between this class and the QueryPlanExecutionStrategy is that
+    it relies on a smarter QueryRunner (SplitQueryRunner) than the one provided to the
+    execution strategy. The runner this class receives is supposed to take care of
+    running the DB query processors before executing the query on the database so that
```

Comment:
I am still a bit nervous of having processors that should be executed right before talking to the database when a splitter is orchestrating this process.
I think, this way, by keeping the responsibility of running the processors out of the splitter:
- we unload some responsibility from the splitter class
- db query processors are entirely a matter of the same class (the execution strategy, none else knows how to run them)
- we constrain the splitter strategy not to be too "creative" 

We do not need the flexibility of letting the splitter tweak which processors are executed, and probably we should try to avoid that. In case there was a need for that, it will be a fairly small scope change.